src/lib.rs

```diff
@@ -46,6 +67,200 @@ impl UniffiCustomTypeConverter for Uuid {
+    mrz_payload: String,
+    qr_payload: String,
+) -> Result<(), VCBVerificationError> {
+    let mrz: MRZ = mrz_payload
+        .lines()
+        .map(|x| *convert_to_mrz_entry(x.as_bytes()))
+        .collect::<Vec<[u8; 30]>>()
+        .try_into()
+        .map_err(|_| VCBVerificationError::Generic {
+            value: "Invalid MRZ string".to_string(),
+        })?;
+
+    // First we decode the QR-code payload to get the VCB in CBOR-LD form.
+    let input = MachineReadableZone::decode_qr_code_payload(qr_payload.as_str()).map_err(|e| {
+        VCBVerificationError::Generic {
+            value: e.to_string(),
+        }
+    })?;
+
+    // Then we decompress the CBOR-LD VCB to get a regular JSON-LD VCB.
+    let vc = decode_from_bytes::<MachineReadableZone>(&input)
+        .await
+        .map_err(|e| VCBVerificationError::Generic {
+            value: e.to_string(),
+        })?;
+
+    // Finally we verify the VCB against the MRZ data.
+    let params = VerificationParameters::new(AnyDidMethod::default().into_vm_resolver());
+    verify(&vc, &mrz, params)
+        .await
+        .map_err(|e| VCBVerificationError::Generic {
+            value: e.to_string(),
+        })?
+        .map_err(|e| VCBVerificationError::Generic {
+            value: e.to_string(),
+        })
+}
+
+#[derive(thiserror::Error, uniffi::Error, Debug)]
+pub enum VCVerificationError {
+    #[error("{value}")]
+    Generic { value: String },
+}
+
+#[uniffi::export]
+pub async fn verify_json_vc_string(json: String) -> Result<(), VCVerificationError> {
+    use ssi::prelude::VerificationParameters;
+
+    let vc = any_credential_from_json_str(&json).map_err(|e| VCVerificationError::Generic {
+        value: e.to_string(),
+    })?;
+
+    let vm_resolver = AnyDidMethod::default().into_vm_resolver();
+    let params = VerificationParameters::from_resolver(vm_resolver);
+
+    vc.verify(&params)
+        .await
+        .map_err(|e| VCVerificationError::Generic {
+            value: e.to_string(),
+        })?
+        .map_err(|e| VCVerificationError::Generic {
+            value: e.to_string(),
+        })
+}
+
+#[derive(thiserror::Error, uniffi::Error, Debug)]
+pub enum VPError {
+    #[error("{value}")]
+    Verification { value: String },
+    #[error("{value}")]
+    Parsing { value: String },
+}
+
+#[uniffi::export]
+pub async fn vc_to_signed_vp(vc: String, key_str: String) -> Result<String, VPError> {
+    use ssi::prelude::*;
+
+    let vp = ssi::claims::vc::v1::JsonPresentation::new(None, None, vec![vc]);
+
+    let mut key: ssi::jwk::JWK = serde_json::from_str(&key_str).map_err(|e| VPError::Parsing {
+        value: e.to_string(),
+    })?;
+    let did = DIDJWK::generate_url(&key.to_public());
+    key.key_id = Some(did.into());
+
+    let jwt = vp
+        .to_jwt_claims()
+        .map_err(|e| VPError::Parsing {
+            value: e.to_string(),
+        })?
+        .sign(&key)
+        .await
+        .map_err(|e| VPError::Parsing {
```

Comment:
This needs a new variant to say that it's the signing that failed, not the parsing